File: Week1/threesum.py
arr = [30, -40, -20, -10, 40, 0, 10, 5]
def create_list_of_numbers(n):
	import random
	return [random.randint(-1000, 10000) for _ in range(n)]


def three_sum(arr):
	sum_of_two = {}
	for i in range(len(arr)):
		for j in range(i+1, len(arr)):
			suum = arr[i] + arr[j]
			if suum in sum_of_two:
				sum_of_two[suum].append((arr[i], arr[j]))
			else:	
				sum_of_two[suum] = [(arr[i], arr[j])]
	
	total_pairs = 0

	sum_of_three = {}
	for i in range(len(arr)):
		if -arr[i] in sum_of_two:
			for pair in sum_of_two[-arr[i]]:
				if arr[i] not in pair:
					triplet = tuple(sorted([arr[i], pair[0], pair[1]]))
					if triplet not in sum_of_three:
						sum_of_three[triplet] = 1
						total_pairs += 1
	return len(sum_of_three)



def three_sum_brute(arr):
	total_pairs = 0
	all_pairs = []
	for i in range(len(arr)):
		for j in range(i+1, len(arr)):
			for k in range(j+1, len(arr)):
				if arr[i] + arr[j] + arr[k] == 0:
					total_pairs += 1
					all_pairs.append(sorted([arr[i], arr[j], arr[k]]))
	 
	return total_pairs


import time
import matplotlib.pyplot as plt
import random

def test_and_plot_three_sum():
	# Sizes and timings
	sizes = list(range(10, 2001, 200))  # Interval of 200
	optimized_times = []
	brute_force_times = []

	for size in sizes:
		print(f"Testing for size {size}")
		arr = list(set(create_list_of_numbers(size)))

		# Measure optimized time
		start_time = time.time()
		three_sum(arr)
		end_time = time.time()
		optimized_times.append(end_time - start_time)

		# Measure brute force time (smaller sizes only to avoid timeout)
		start_time = time.time()
		three_sum_brute(arr)
		end_time = time.time()
		brute_force_times.append(end_time - start_time)

	print(sizes)
	print(optimized_times)
	print(brute_force_times)
	# Plot results
	plt.figure(figsize=(12, 6))
	plt.plot(sizes, optimized_times, label="Optimized Approach", marker='o')

	# Plot brute force times only for available data
 
	
	brute_sizes = [sizes[i] for i in range(len(sizes)) if brute_force_times[i] is not None]
	brute_times = [time for time in brute_force_times if time is not None]
	plt.plot(brute_sizes, brute_times, label="Brute Force Approach", marker='x' )

	plt.xlabel("Input Size (n)")
	plt.ylabel("Time Taken (seconds)")
	plt.title("Three Sum Performance: Optimized vs Brute Force")
	plt.legend()
	plt.grid()
	plt.show()


test_and_plot_three_sum()

File: Week1/threesum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <set>
#include <cstdlib>
#include <ctime>
#include <algorithm>

// Function to create a list of random numbers
std::vector<int> create_list_of_numbers(int n) {
    std::vector<int> arr;
    for (int i = 0; i < n; ++i) {
        arr.push_back(rand() % 11001 - 1000); // Random number between -1000 and 10000
    }
    return arr;
}

// Optimized three_sum function
int three_sum(const std::vector<int>& arr) {
    std::unordered_map<int, std::vector<std::pair<int, int>>> sum_of_two;
    for (size_t i = 0; i < arr.size(); ++i) {
        for (size_t j = i + 1; j < arr.size(); ++j) {
            int suum = arr[i] + arr[j];
            sum_of_two[suum].emplace_back(arr[i], arr[j]);
        }
    }

    std::set<std::tuple<int, int, int>> sum_of_three;
    for (size_t i = 0; i < arr.size(); ++i) {
        int target = -arr[i];
        if (sum_of_two.find(target) != sum_of_two.end()) {
            for (const auto& pair : sum_of_two[target]) {
                if (arr[i] != pair.first && arr[i] != pair.second) {
                    std::vector<int> triplet = {arr[i], pair.first, pair.second};
                    std::sort(triplet.begin(), triplet.end());
                    sum_of_three.emplace(triplet[0], triplet[1], triplet[2]);
                }
            }
        }
    }

    return sum_of_three.size();
}

// Brute force three_sum function
int three_sum_brute(const std::vector<int>& arr) {
    int total_pairs = 0;
    std::set<std::vector<int>> all_pairs;
    for (size_t i = 0; i < arr.size(); ++i) {
        for (size_t j = i + 1; j < arr.size(); ++j) {
            for (size_t k = j + 1; k < arr.size(); ++k) {
                if (arr[i] + arr[j] + arr[k] == 0) {
                    std::vector<int> triplet = {arr[i], arr[j], arr[k]};
                    std::sort(triplet.begin(), triplet.end());
                    if (all_pairs.find(triplet) == all_pairs.end()) {
                        all_pairs.insert(triplet);
                        ++total_pairs;
                    }
                }
            }
        }
    }

    return total_pairs;
}

// Test and print results
void test_and_print_three_sum() {
    std::srand(std::time(0)); // Seed random number generator
    std::vector<int> sizes;
    for (int n = 10; n <= 5000; n += 200) {
        sizes.push_back(n);
    }

    std::vector<double> optimized_times;
    std::vector<double> brute_force_times;

    for (int size : sizes) {
        std::vector<int> arr = create_list_of_numbers(size);

        // Measure optimized time
        clock_t start_time = clock();
        int num_pair_opt = three_sum(arr);
        clock_t end_time = clock();
        optimized_times.push_back(double(end_time - start_time) / CLOCKS_PER_SEC);

        // Measure brute force time (for small sizes only to avoid timeout)
        start_time = clock();
        int num_pair_brute =  three_sum_brute(arr);
        end_time = clock();
        brute_force_times.push_back(double(end_time - start_time) / CLOCKS_PER_SEC);

        std::cout << "Size: " << size
                  << " | Optimized Time: " << optimized_times.back()
                  << "s | Brute Force Time: " << brute_force_times.back()
				  << "s | Optimized Pairs: " << num_pair_opt
				  << " | Brute Force Pairs: " << num_pair_brute
				  << std::endl;
    }
}

int main() {
    test_and_print_three_sum();
    return 0;
}
